### ISSUE TITLE
Twisted shouldn't start processes until the reactor is running

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2078]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#2078|
|Type|enhancement|
|Created|2006-09-11 22:40:25Z|
|Branch|https://github.com/twisted/twisted/tree/eliminate-zombie-warning-2078|

You get warnings and maybe bad stuff happens because the sigchild handler hasn't been installed until the reactor is started.

Attachments:

* [zombie-test.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bcc/bcc3fe1f6a2e6fe2a723905628738169ff3eca81/13466e6e941b0f6d30aa3ffc77abaca22a945b4f.patch) (6400 bytes) - added by kelly on 2009-11-05 04:57:25Z - Adds a test that demonstrates a zombie under the current system. Removes the other PotentialZombieWarning tests.
* [zombie-fix.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bcc/bcc3fe1f6a2e6fe2a723905628738169ff3eca81/65eddc67752ae7cea9c1a27f54eebb7c461ab7c2.patch) (457 bytes) - added by kelly on 2009-11-05 04:59:54Z - Runs the SIGCHLD handler immediately after installing it.
* [zombie-deprecate.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bcc/bcc3fe1f6a2e6fe2a723905628738169ff3eca81/9b307c77b0fc53418098cd4b62b87597ba4382f2.patch) (19951 bytes) - added by kelly on 2009-11-05 05:01:13Z - Removes all use of PotentialZombieWarning and deprecates it.
* [whenRunning.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bcc/bcc3fe1f6a2e6fe2a723905628738169ff3eca81/caca9e23a9369d9fdf038c8bcea0b13fdbb1c184.patch) (6404 bytes) - added by kelly on 2009-11-05 05:03:22Z - A different way of achieving the same goal.  Adds a decorator for tests that causes them to run after the reactor has started.

<details><summary>Searchable metadata</summary>

```
trac-id__2078 2078
type__enhancement enhancement
reporter__radix radix
priority__normal normal
milestone__ 
branch__branches_eliminate_zombie_warning_2078 branches/eliminate-zombie-warning-2078
branch_author__kelly__exarkun kelly, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1158014425000000 1158014425000000
changetime__1261149993000000 1261149993000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

This probably hasn't been resolved yet because it'd be tricky to implement for `IReactorProcess.spawnProcess`.  The returned `IProcess` is currently immediately valid.  Delaying process spawning would involve sometimes returning one which could not be used yet.  This is a nice idea, and I think the `spawnProcess` replacement should consider working this way, but it probably isn't workable for `spawnProcess`.

As an alternative, I propose that we drop `PotentialZombieWarning` entirely and just call the appropriate reap process function when the reactor is started (more specifically, when it installs the SIGCHLD handler).  This will clean up any processes which were started _and_ exited before the signal handler was installed.  No zombies.  Hooray.

**Comment by twisted-trac** [comment]
|<img alt="kelly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|kelly commented|
|-|-|

Replying to [exarkun](#note_2):
> As an alternative, I propose that we drop `PotentialZombieWarning`

Just for fun, the attached series of three patches implements exarkun's suggestion (a 12 line patch supported by 830 lines for tests).

My original approach to suppressing `PotentialZombieWarning` in the test suite is contained in the fourth (unrelated patch), for reference :-)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Can you describe a little more clearly what is up for review here?  There seem to be an awful lot of patches on this ticket now.  Ideally, there should be just one.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to kelly|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kelly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|kelly set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_4):
> Can you describe a little more clearly what is up for review here?  There seem to be an awful lot of patches on this ticket now.  Ideally, there should be just one.

Sorry - My idea was that breaking it into three parts would make it easier to review.

The parts up for review are:

 * [/trac/attachment/ticket/2078/zombie-test.2.patch zombie-test.2.patch]

This patch replaces the PotentialZombieWarning tests with a test that demonstrates a zombie process.  Once applied against trunk the test is expected to fail.

 * [/trac/attachment/ticket/2078/zombie-fix.patch zombie-fix.patch]

This patch fixes the problem as you suggested in [#comment:2 an earlier comment].  Specifically, it runs the SIGCHLD handler immediately after installing it.

 * [/trac/attachment/ticket/zombie-deprecate.patch zombie-deprecate.patch]

This patch removes all usage of PotentialZombieWarning in twisted and deprecates the class.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to kelly|
|-|-|

Thanks, that helps me a bunch. :)  Having the patches separate may actually make the review easier, once I know what each is supposed to do.

  1. First, `twisted.internet.test.test_process.ProcessTestsBuilder_Glib2Reactor.test_spawnProcessEarlyIsReaped` fails for me. :/  Ubuntu Karmic, if that helps any.  Somehow, several other apparently unrelated tests also fail with these patches applied: `twisted.test.test_tcp.LoopbackTestCase.test_connectionRefusedErrorNumber`, `twisted.test.test_tcp.ConnectorTestCase.test_reconnect`, `twisted.test.test_tcp.ListeningTestCase.test_directConnectionLostCall`, `twisted.test.test_tcp.LoopbackTestCase.testFailing`.  I can provide more details if you aren't able to reproduce this.  I suspect their failure is related to what the new `test_error.py` does, though.
  1. Is the new `foo = ` in `test_process.py` doing anything, or is it an accident/left-over that can be un-added?
  1. I worry about the timeouts in `test_spawnProcessEarlyIsReaped`.  I wonder if `signal.pause` could be used here.  Unfortunately it doesn't accept a timeout.  One can cobble a timeout together with `signal.alarm`.  Then suddenly things are getting complicated again, though.  Either way, waiting for longer than a second is a good idea.  I suggest at least 60 seconds.  The default timeout applied to trial tests is 120 seconds.
  1. The context-manager approach is optimistic. :)  It will be a long time before we're using `with`, though.  Can you just write this the straightforward way with try/finally?
  1. You shouldn't need to check `reactor.running`.  What case does this check help out in?
  1. In `twisted/internet/error.py`, the `warnings` module is imported but unused.
  1. Exceptions aren't new-style in older versions of Python, so you can't use `super` to call the `PotentialZombieWarning` base's `__init__`.
  1. In `twisted/internet/process.py`, `_BaseProcess.__init__` looks like it doesn't need to exist at all anymore.
  1. The new `test_error.py` needs a copyright statement and a module docstring.  Also, several of the names it uses are in the wrong style.  Please use `camelCase`.
  1. Most of the tests in this module are weird.  What motivated you to write `GeneralExceptionTests`?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [27753]) Branching to 'eliminate-zombie-warning-2078'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [27754]) Apply kelly's patches removing the need for PotentialZombieWarning

refs [#2078](https://github.com/twisted/twisted/issues/2078)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner|
|-|-|

Review points addressed.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/eliminate-zombie-warning-2078).  Please review.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|TimAllen set owner to @exarkun|
|-|-|

1.In `twisted/internet/test/reactormixins.py` there's an added docstring that says "A C{list} of interfaces which the require must provide" - "require" is probably a tab-complete-o for "reactor".
 1. In `twisted/internet/test/test_process.py`, in `ProcessTestsBuilderBase.test_spawnProcessEarlyIsReaped`, the variable `childFDs` is set to None, then set again if `self.usePTY` is true. This seems redundant.

Apart from that, all looks good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [27785]) Address review comments

refs [#2078](https://github.com/twisted/twisted/issues/2078)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

I tried the branch on all the builders once more and one of them had a failure of the new test. :/  Looks like there's a race in it, between `registerReapHandler` and getting `SIGCHLD`.  If `SIGCHLD` arrives first, `registerReapHandler` synchronously calls `processEnded` which causes the `Deferred` in the test to fire too soon.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [27786]) Avoid process ended callbacks before the reactor has started

This fixes intermitent failures in the new test caused by the registerReapProcessHandler/SIGCHLD delivery race
going in favor of the SIGCHLD delivery.

refs [#2078](https://github.com/twisted/twisted/issues/2078)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun removed owner|
|-|-|

Here's the failure I saw on buildbot:

```
===============================================================================
[FAIL]: twisted.internet.test.test_process.ProcessTestsBuilder_SelectReactor.test_spawnProcessEarlyIsReaped

Traceback (most recent call last):
  File "/home/buildslave/slaves/twisted/full2.4-debian64/Twisted/twisted/internet/test/test_process.py", line 87, in test_spawnProcessEarlyIsReaped
    self.assertFalse(result)
twisted.trial.unittest.FailTest: None
-------------------------------------------------------------------------------
```

Here's the latest [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/eliminate-zombie-warning-2078).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @exarkun|
|-|-|

So much for that idea.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

Okay, I just accounted for the synchronous case in the new unit test.  This is a better idea anyway, since this behavior existed _before_ this change; I shouldn't be trying to change it here, since it's not directly relevant to the zombie warning.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [27787]) Get rid of that behavior change and just handle the current behavior in the new test

refs [#2078](https://github.com/twisted/twisted/issues/2078)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun removed owner|
|-|-|

New build results in the same place.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve set owner to @exarkun|
|-|-|

[1]
```
+        # processEnded synchronously, triggering an application-visible
+        # callback.  That's probably not idea.  The replacement API for
```

Typo: ideal.

[2]

```
+            self._handleSigchld(signal.SIGCHLD, None)
```

Can you add a comment here?

Great change, +1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [27820]) fix typo, add comment about the direct SIGCHLD handler call

refs [#2078](https://github.com/twisted/twisted/issues/2078)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set status to `closed`|
|-|-|

(In [27824]) Merge eliminate-zombie-warning-2078

Author: kelly, exarkun
Reviewer: TimAllen, therve
Fixes: [#2078](https://github.com/twisted/twisted/issues/2078)

Check for exited child processes when the reactor starts, eliminating the possibility
of ever-lasting zombies created by reactor.spawnProcess calls before a call to
reactor.run.

`PotentialZombieWarning` is also deprecated, now that it can never be emitted.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
